blueocean-dashboard/src/main/js/util/dedupe-calls.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * DuplicateCallTracker maintains active calls against a particular key
+ */
+export class DeDupeCallTracker {
+    constructor() {
+        /**
+         * Onload callbacks cache. Used to ensure we don't
+         * issue multiple in-parallel requests for the same
+         * class metadata.
+         */
+        this.promises = {};
+    }
+
+    /**
+     * Generalization of duplicate request consolidation:
+     *
+     * @key: key to use to track the duplicate requests
+     * @promiseCreator: function that will return an initial promise, e.g. () => fetch(...)
+     * @return a Promise
+     */
+    dedupe(key, promiseCreator) {
+        // get active or create
+        return this.promises[key] || (this.promises[key] =
+            promiseCreator()
+            .then((data) => {
+                delete this.promises[key];
+                return data;
+            })
+            .catch((err) => {
+                delete this.promises[key];
+                return Promise.reject(err);
```

Comment:
Per https://github.com/matthew-andrews/isomorphic-fetch we should probably add the "es6-promise"; just jumped out at me because it's first time I saw a ref to native Promise in our code.